## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [206]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [207]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [208]:
# Use the describe function to gather some basic statistics
print(df.describe)

<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

In [209]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.head(30)

first        last                                           job  \
0        Donald     Harding                                  Immunologist   
1       Tiffany       Myers                               Music therapist   
2       Shannon      Watson                                  Immunologist   
3        Nathan       Baker                         Accounting technician   
4     Christina    Schwartz                                  Chiropractor   
5       Jessica       Reyes  Production designer, theatre/television/film   
6         Brian       Moore                                       Barista   
7         Jesus      Wright                            Theme park manager   
8       Rebecca       Brown                                   Chiropodist   
9          Greg      Garcia                            Secretary, company   
10         Kyle    Williams           Engineer, broadcasting (operations)   
11      Roberto      Howard                      Surveyor, land/geomatics   
12      Jessica       Reyes  Production designer, theatre/television/film   
13     Patricia       Gould                                      Lobbyist   
14    Katherine      Taylor                        Psychologist, clinical   
15      Matthew      Orozco                    Research scientist (maths)   
16       Joseph      Jordan            Government social research officer   
17       Carrie       Green                           Colour technologist   
18        David       Reyes              Production assistant, television   
19       Shelby    Williams                           Colour technologist   
20       Tricia  Stephenson                   Lecturer, further education   
21      Patrick    Anderson                Commercial art gallery manager   
22        Peter     Roberts                    Architectural technologist   
23  Christopher      Rhodes               Engineer, manufacturing systems   
24        Glenn      Garcia                              Textile designer   
25        Peter     Roberts                    Architectural technologist   
26      Patrick    Anderson                Commercial art gallery manager   
27        Jacob     Mendoza                                Air cabin crew   
28      Tiffany        Lowe                           Medical illustrator   
29        Caleb        Rice                              Technical brewer   

                     phone                                     email  \
0       793-904-7725x39308          harding.donald.7185@sullivan.com   
1         201.442.4543x942         myers.t.6537@ferguson-johnson.net   
2        687.737.9424x8503                     swatson8146@payne.net   
3         827-788-8123x012                    bakernathan@benson.com   
4             265-829-3643       christinaschwartz9252@mcconnell.com   
5               5016915923                  jessica.reyes@grimes.net   
6      +1-893-400-2733x982               moore.official@martinez.com   
7       605.947.9467x66513             wright.jesus.7579@meadows.com   
8         665-955-0156x926                      rbrown850@lewis.info   
9       (231)342-7216x1421                       ggarcia@bentley.biz   
10         +1-432-234-2942  communitywilliams9024@taylor-padilla.biz   
11            311.284.4575      roberto.howard.1290@davis-reese.info   
12              5016915923                  jessica.reyes@grimes.net   
13   +1-694-749-5807x94968                  newspapergould@walsh.biz   
14       (943)632-3743x028         taylor.katherine.733@odonnell.org   
15    +1-259-213-7449x4333                    orozcomatthew@carr.com   
16      680.657.3271x16477             jordan_joseph_4684@gibson.net   
17     +1-242-699-5329x474                  c.green@oliver-smith.com   
18            714.475.3974                    davidreyes@daniels.org   
19        664.584.8160x389               williamstry70@armstrong.com   
20       (978)811-8729x522              triciastephenson@vasquez.org   
21      947-812-9604x27025       ander

In [210]:
# What three item categories had the most entries?
category_item_count = df["category"].value_counts()
category_item_count.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [211]:
# For the category with the most entries, which subcategory had the most entries?
subcategory_find = df[df["category"] == "consumables"]
subcategory_count = subcategory_find['subcategory'].value_counts()

most_subcategory = subcategory_count.idxmax
most_subcategory

<bound method Series.idxmax of subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64>

In [212]:
# Which five clients had the most entries in the data?
client_count = df["client_id"].value_counts()

top_client_count = client_count.head(5)
top_client_count

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [213]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_client_count.index.tolist()

In [214]:
# How many total units (the qty column) did the client with the most entries order order?
most_buying_client = top_client_count.index[0]
client_find = df[df["client_id"] == most_buying_client]
order_amount = client_find["qty"].sum()
order_amount

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [215]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["subtotal"] = df["unit_price"] * df["qty"]


In [216]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["shipping_price"] = df["unit_weight"] * df["qty"]
df["shipping_price"] = df["shipping_price"].apply(lambda x: 7 if x > 50 else 10)
df["shipping_price"] = df["shipping_price"].apply(lambda x: round(x, 2))



In [217]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = df["subtotal"] + df["shipping_price"]
df["total_price"] = df["total_price"] + (df["total_price"] * 0.0925)
df["total_price"] = df["total_price"].apply(lambda x: round(x, 2))


In [218]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["cost"] = df["unit_cost"] * df["qty"] + df["shipping_price"]
df["cost"] = df["cost"].apply(lambda x: round(x, 2))

In [219]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = df["total_price"] - df["cost"]
df.head(30)


first        last                                           job  \
0        Donald     Harding                                  Immunologist   
1       Tiffany       Myers                               Music therapist   
2       Shannon      Watson                                  Immunologist   
3        Nathan       Baker                         Accounting technician   
4     Christina    Schwartz                                  Chiropractor   
5       Jessica       Reyes  Production designer, theatre/television/film   
6         Brian       Moore                                       Barista   
7         Jesus      Wright                            Theme park manager   
8       Rebecca       Brown                                   Chiropodist   
9          Greg      Garcia                            Secretary, company   
10         Kyle    Williams           Engineer, broadcasting (operations)   
11      Roberto      Howard                      Surveyor, land/geomatics   
12      Jessica       Reyes  Production designer, theatre/television/film   
13     Patricia       Gould                                      Lobbyist   
14    Katherine      Taylor                        Psychologist, clinical   
15      Matthew      Orozco                    Research scientist (maths)   
16       Joseph      Jordan            Government social research officer   
17       Carrie       Green                           Colour technologist   
18        David       Reyes              Production assistant, television   
19       Shelby    Williams                           Colour technologist   
20       Tricia  Stephenson                   Lecturer, further education   
21      Patrick    Anderson                Commercial art gallery manager   
22        Peter     Roberts                    Architectural technologist   
23  Christopher      Rhodes               Engineer, manufacturing systems   
24        Glenn      Garcia                              Textile designer   
25        Peter     Roberts                    Architectural technologist   
26      Patrick    Anderson                Commercial art gallery manager   
27        Jacob     Mendoza                                Air cabin crew   
28      Tiffany        Lowe                           Medical illustrator   
29        Caleb        Rice                              Technical brewer   

                     phone                                     email  \
0       793-904-7725x39308          harding.donald.7185@sullivan.com   
1         201.442.4543x942         myers.t.6537@ferguson-johnson.net   
2        687.737.9424x8503                     swatson8146@payne.net   
3         827-788-8123x012                    bakernathan@benson.com   
4             265-829-3643       christinaschwartz9252@mcconnell.com   
5               5016915923                  jessica.reyes@grimes.net   
6      +1-893-400-2733x982               moore.official@martinez.com   
7       605.947.9467x66513             wright.jesus.7579@meadows.com   
8         665-955-0156x926                      rbrown850@lewis.info   
9       (231)342-7216x1421                       ggarcia@bentley.biz   
10         +1-432-234-2942  communitywilliams9024@taylor-padilla.biz   
11            311.284.4575      roberto.howard.1290@davis-reese.info   
12              5016915923                  jessica.reyes@grimes.net   
13   +1-694-749-5807x94968                  newspapergould@walsh.biz   
14       (943)632-3743x028         taylor.katherine.733@odonnell.org   
15    +1-259-213-7449x4333                    orozcomatthew@carr.com   
16      680.657.3271x16477             jordan_joseph_4684@gibson.net   
17     +1-242-699-5329x474                  c.green@oliver-smith.com   
18            714.475.3974                    davidreyes@daniels.org   
19        664.584.8160x389               williamstry70@armstrong.com   
20       (978)811-8729x522              triciastephenson@vasquez.org   
21      947-812-9604x27025       ander

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [221]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_client_spend = df.groupby("client_id")["total_price"].sum()
top_client_spend = top_client_spend.loc[top_client_ids]
top_client_spend


client_id
33615     6380943.04
66037     8737039.74
46820     7995940.84
38378     9161581.67
24741    76669977.65
Name: total_price, dtype: float64

In [222]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
top_clients_df = df[df['client_id'].isin(top_client_ids)]
top_clients_totals = top_clients_df.groupby('client_id').agg(
    total_units=('qty', 'sum'),
    total_shipping=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
)
top_clients_df.head()

first     last                                           job  \
5    Jessica    Reyes  Production designer, theatre/television/film   
12   Jessica    Reyes  Production designer, theatre/television/film   
137    Bryan    Myers                                Chief of Staff   
214    Bryan    Myers                                Chief of Staff   
216   Angela  Everett                                     Osteopath   

                phone                     email  client_id  order_id  \
5          5016915923  jessica.reyes@grimes.net      33615   1327730   
12         5016915923  jessica.reyes@grimes.net      33615   6059361   
137  444.233.9146x233     bmyers3696@wilson.com      46820   7109502   
214  444.233.9146x233     bmyers3696@wilson.com      46820   8967498   
216  729-620-6192x875    aeverett@rodriguez.com      66037   1203693   

     order_date  order_week  order_year  ... unit_price unit_cost unit_weight  \
5    2023-05-02          18        2023  ...     172.84    154.44        2.79   
12   2023-05-16          20        2023  ...     229.74    198.38       27.51   
137  2023-05-12          19        2023  ...     136.13    111.47        5.65   
214  2023-02-06           6        2023  ...       7.98      6.26        7.05   
216  2023-04-09          14        2023  ...     301.87    166.09        0.26   

     qty  line_number   subtotal  shipping_price  total_price       cost  \
5     78            0   13481.52               7     14736.21   12053.32   
12   568            2  130492.32               7    142570.51  112686.84   
137  118            9   16063.34               7     17556.85   13160.46   
214  133            3    1061.34               7      1167.16     839.58   
216  157            2   47393.59              10     51788.42   26086.13   

       profit  
5     2682.89  
12   29883.67  
137   4396.39  
214    327.58  
216  25702.29  

[5 rows x 23 columns]

In [223]:
top_clients_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'subtotal', 'shipping_price', 'total_price', 'cost',
       'profit'],
      dtype='object')

In [224]:
# Format the data and rename the columns to names suitable for presentation.
top_clients_totals.rename(columns={
    'total_units': 'Total Units Purchased',
    'total_shipping': 'Total Shipping Price',
    'total_revenue': 'Total Revenue',
    'total_profit': 'Total Profit'
}, inplace=True)

# Define the money columns. 
money_columns = ['Total Shipping Price', 'Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return f"${value / 1000000:.2f}M"
# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    top_clients_totals[col] = top_clients_totals[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top_clients_totals.rename(columns={
    'Total Shipping Price': 'Total Shipping Price (Millions)',
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Profit': 'Total Profit (Millions)'
}, inplace=True)


In [225]:
top_clients_totals.columns

Index(['Total Units Purchased', 'Total Shipping Price (Millions)',
       'Total Revenue (Millions)', 'Total Profit (Millions)'],
      dtype='object')

In [226]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
top_clients_totals = top_clients_totals.sort_values('Total Profit (Millions)', ascending=False)
top_clients_totals.head()


Total Units Purchased Total Shipping Price (Millions)  \
client_id                                                          
24741                     239862                          $0.00M   
66037                      43018                          $0.00M   
38378                      73667                          $0.00M   
46820                      75768                          $0.00M   
33615                      64313                          $0.00M   

          Total Revenue (Millions) Total Profit (Millions)  
client_id                                                   
24741                      $76.67M                 $36.11M  
66037                       $8.74M                  $3.13M  
38378                       $9.16M                  $2.95M  
46820                       $8.00M                  $2.59M  
33615                       $6.38M                  $2.03M